core/src/jsMain/kotlin/Characteristic.kt

```diff
@@ -0,0 +1,26 @@
+package com.juul.kable
+
+import com.benasher44.uuid.Uuid
+import com.juul.kable.external.BluetoothRemoteGATTCharacteristic
+import com.juul.kable.external.BluetoothServiceUUID
+
+public actual class Characteristic internal constructor(
+    internal val serviceUuid: Uuid,
+    internal val bluetoothRemoteGATTCharacteristic: BluetoothRemoteGATTCharacteristic,
+) {
+
+    public actual val uuid: Uuid by lazy { bluetoothRemoteGATTCharacteristic.uuid.toUuid() }
+
+    public actual val descriptors: List<Descriptor>
+        get() = TODO()
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class.js != other::class.js) return false
+        other as Characteristic
+        if (uuid != other.uuid) return false
+        return true
+    }
+
+    override fun hashCode(): Int = uuid.hashCode()
```

Comment:
Just checking. Is it correct that we're only using the `uuid` for equality?

Reply:
Being that these are on the public API (and sadly the `data` `ByteArray` is mutable, thanks Java), then no, it should be more than just UUID. I'll think about the best way to fix this (I'm hoping I can think of a clever way to not have to incur the expense of comparing the `ByteArray` contents.

Reply:
**EDIT**: I misread the code this was referring to.

The equality check really only needs to cover what is public, so, in this case it is still incorrect (as `descriptors` is public), but it won't be as expensive as I originally assumed (we aren't doing a content equality check of an arbitrary length `ByteArray`).

Granted we also won't know how many potential descriptors, but in most cases those will be small, and comparing them will involve comparing their UUIDs.

Reply:
Yikes. Thought about this further. I can see where this could really bite someone.

I'm using `Characteristic` (and friends) to hold a reference to the actual platform characteristic. This is a handy "optimization" as anytime the consumer of the library invokes any I/O function they do it with an object that has the direct reference to the type needed to execute the I/O operation.

Unfortunately, that also means that if the connection drops and is re-established, then this `Characteristic` will likely not work (worse yet, I believe on Android it would still work, but not on JS — inconsistent behavior I definitely want to avoid).

The alternative is the have `Characteristic` (and friends) be basic models (just holding the needed UUIDs). This would greatly simplify these classes (as no need to override `equals`/`hashCode`) **but** would mean that every I/O operation would require a lookup of the actual platform object to perform the I/O operation.

I believe the expense would be: O(n) + O(n) + O(n) = O(n) — a list search for service UUID, characteristic UUID and potentially descriptor UUID. All the `n`s would likely be small, but an expense nonetheless.

Although as @cedrickcooke is always quick to remind me: this is I/O, these tasks are likely significantly smaller than actually performing the I/O.

Reply:
Was surprised to hear my PR comments uncovered anything major 😆  they all were surface level when I wrote them.

Not sure how difficult it would be/if even possible, but can you go for the best of both worlds? Characteristics carry what they need for a fast happy path, and if the connection drops and is re-established it will take the slow route to recalculate everything?

Reply:
Actually, I was thinking of something along those lines: lazy characteristics.

Essentially, you create a lazy `Characteristic` via a function on `Peripheral` (e.g. `Peripheral.characteristic(Uuid, Uuid)`). It would simply hold the UUID for the service and characteristic. When needed, the `Peripheral` object would determine that it was lazy and resolve the actual platform object (and store it on the lazy characteristic object).

On disconnect, would need to invalidate all lazy characteristics so that they would again be in their initial lazy state. This is where the hard part comes in, having an efficient way to invalidate.

Was thinking of having the peripheral keep a list of all "child" lazy characteristics, but that sounds like it could be error prone. As consumers of the library make create lots of characteristic objects not realize that a reference is being held to them.

Alternatively, could have an internal monotonic sequence ID. Then when connection drops mark what the sequence ID was at, any subsequent I/O requests with a characteristic with an "old" sequence ID would know to be re-resolved.

Reply:
Lazy characteristics also carry a huge advantage: you can set them up **before** a connection has been established.

This is a huge win, as setting up everything (including observations) but having to wait until **after** connected (and services discovered) is a huge pain.

It'd be amazing to have an API that allowed an entire peripheral representation (e.g. SensorTag) to be setup ahead of time. You could even setup the `observe` call before connected. When connected the library would take care of getting everything hooked up.

The challenging part will be ensuring that stacktraces (when things go wrong) are helpful. With non-lazy characteristics your stacktrace would be clear that your call to retrieve a non-existent characteristic originated from a specific part of your code. With lazy characteristics, the characteristic resolution will happen automatically so be a bit removed from where the characteristic was setup.

I could have a debug mode (similar to Coroutines), when enabled, it would build a stracktrace for the creation of the characteristic and store that. It could be thrown as the cause of any failures to make it clear where to creation originated.

Reply:
Does Kotlin Multiplatform have an equivalent of `WeakReference`?

Reply:
Love the sequence ID idea. Seems easy enough to implement and cheap enough at runtime.